### PR TITLE
[Pie Chart] Add three new attributes Half circle,Pad Angle,Corner Radius to Pie Chart

### PR DESCRIPTION
**Pull request by ganshanshan**
Add three new attributes Half circle,Pad Angle,Corner Radius to Pie Chart

### CODE REVIEW

File: superset/viz.py [hunk 1]
```diff
@@ -1207,16 +1207,34 @@ class DistributionPieViz(NVD3Viz):
     viz_type = "pie"
     verbose_name = _("Distribution - NVD3 - Pie Chart")
     is_timeseries = False
+    # fieldsets = ({
```
Comment (xrmx) [inline superset/viz.py]: no commented out code please

File: superset/assets/visualizations/nvd3_vis.js [hunk 2]
```diff
@@ -195,6 +195,33 @@ function nvd3Vis(slice) {
...
             chart.cornerRadius(true);
+
+            // Configure the pie more tyles
+            if (fd.pie_half_circle){
+              chart.pie
+              .startAngle(function(d) { return d.startAngle/2 -Math.PI/2 })
+              .endAngle(function(d) { return d.endAngle/2 -Math.PI/2 });
+            } 
+
+            if (fd.pie_title){
```
Comment (xrmx) [inline superset/assets/visualizations/nvd3_vis.js]: I don't see references to pie_title, pie_title_pffset and pie_pad_angle anywhere else. Am i missing something?

File: superset/forms.py [hunk 3]
```diff
@@ -981,6 +981,48 @@ def __init__(self, viz):
                 ],
                 "description": _("The color for points and clusters in RGB")
             }),
+            'pie_pad_angle': (FreeFormSelectField, {
+                "label": _("Pad Angle"),
+                "default": 0,
+                "choices": self.choicify([
+                    0,
```
Comment (xrmx) [inline superset/forms.py]: one line should be enough for them

File: superset/forms.py [hunk 4]
```diff
@@ -981,6 +981,48 @@ def __init__(self, viz):
...
+                ]),
+                "description": _("Pad Angle of default viewport."
+                    "Between 0 and 0.1." )
+                
+            }),
+            'pie_corner_radius': (FreeFormSelectField, {
+                "label": _("Corner Radius"),
+                "default": 0,
+                "choices": self.choicify([
+                    0,
```
Comment (xrmx) [inline superset/forms.py]: same here

File: superset/assets/visualizations/nvd3_vis.js [hunk 5]
```diff
@@ -195,6 +195,27 @@ function nvd3Vis(slice) {
             chart.labelThreshold(0.05)  // Configure the minimum slice size for labels to show up
               .labelType(fd.pie_label_type);
             chart.cornerRadius(true);
+            if (fd.pie_half_circle){ // Configure the pie more tyles
+              chart.pie
+              .startAngle(function(d) { return d.startAngle/2 -Math.PI/2 })
+              .endAngle(function(d) { return d.endAngle/2 -Math.PI/2 });
+            } 
+            if (fd.pie_title){
```
Comment (xrmx) [inline superset/assets/visualizations/nvd3_vis.js]: I don't see references of pie_title, pie_title_offset anywhere else.

**Comment by ganshanshan** [comment]
@xrmx thank you for your advice.I correct it now, you can have a check again.
Look forward to your reply.

**Comment by xrmx** [comment]
@ganshanshan you ignored half of my comments

File: superset/forms.py [hunk 6]
```diff
@@ -981,6 +981,46 @@ def __init__(self, viz):
                 ],
                 "description": _("The color for points and clusters in RGB")
             }),
+            'pie_pad_angle': (FreeFormSelectField, {
+                "label": _("Pad Angle"),
+                "default": 0,
+                "choices": self.choicify([
+                    0,
```
Comment (xrmx) [inline superset/forms.py]: one line for all of them should be enough

File: superset/forms.py [hunk 7]
```diff
@@ -981,6 +981,46 @@ def __init__(self, viz):
...
+                    0.09,
+                ]),
+                "description": _("Pad Angle of default viewport."
+                                 "Between 0 and 0.1.")
+            }),
+            'pie_corner_radius': (FreeFormSelectField, {
+                "label": _("Corner Radius"),
+                "default": 0,
+                "choices": self.choicify([
+                    0,
```
Comment (xrmx) [inline superset/forms.py]: same here

File: superset/assets/visualizations/nvd3_vis.js [hunk 8]
```diff
@@ -195,6 +195,27 @@ function nvd3Vis(slice) {
...
+            if (fd.pie_half_circle) { // Configure the pie more tyles
+              chart.pie
+                .startAngle(function (d) {
+                  return d.startAngle / 2 - Math.PI / 2;
+                })
+                .endAngle(function (d) {
+                  return d.endAngle / 2 - Math.PI / 2;
+                });
+            }
+            if (fd.pie_title) {
```
Comment (xrmx) [inline superset/assets/visualizations/nvd3_vis.js]: where is pie_title set?

File: superset/assets/visualizations/nvd3_vis.js [hunk 9]
```diff
@@ -195,6 +195,27 @@ function nvd3Vis(slice) {
...
+              chart.title(fd.pie_title);
+            }
+            if (fd.pie_title_offset) {
```
Comment (xrmx) [inline superset/assets/visualizations/nvd3_vis.js]: same here

File: superset/assets/visualizations/nvd3_vis.js [hunk 10]
```diff
@@ -195,6 +195,27 @@ function nvd3Vis(slice) {
...
+              chart.titleOffset(fd.pie_title_offset);
+            }
+            if (fd.pie_pad_angle) {
```
Comment (xrmx) [inline superset/assets/visualizations/nvd3_vis.js]: if pie_pad_angle default is 0 you don't need the if

**Comment by ganshanshan** [comment]
@xrmx Sorry for my mistake,I have update it.You can have a check.

**Comment by ganshanshan** [comment]
Hi @xrmx ,if you think my code is okay,could you please merge pull reqiest?
Thanks again.

File: superset/models.py [hunk 11]
```diff
@@ -1149,6 +1149,9 @@ def visit_column(element, compiler, **kw):
             qry.compile(
                 engine, compile_kwargs={"literal_binds": True},),
         )
+        template_processor = get_template_processor(
```
Comment (xrmx) [inline superset/models.py]: looks unrelated

File: superset/forms.py [hunk 12]
```diff
@@ -981,6 +981,27 @@ def __init__(self, viz):
...
+                "choices": self.choicify(
+                    [0, 0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09]),
+                "description": _("Pad Angle of default viewport."
+                                 "Between 0 and 0.1.")
+            }),
+            'pie_corner_radius': (FreeFormSelectField, {
+                "label": _("Corner Radius"),
+                "default": 0,
+                "choices": self.choicify([0, 1, 2, 3, 4, 5, 6, 7, 8]),
+                "description": _("Pad Angle of default viewport."
```
Comment (xrmx) [inline superset/forms.py]: looks like it's  between 0 and 8

File: superset/viz.py [hunk 13]
```diff
@@ -1208,14 +1208,21 @@ class DistributionPieViz(NVD3Viz):
     verbose_name = _("Distribution - NVD3 - Pie Chart")
     is_timeseries = False
     fieldsets = ({
-        'label': None,
+        'label': 'Criterion',
```
Comment (xrmx) [inline superset/viz.py]: Do we really need the label here? Criterion does not look like the right one anyway.

File: superset/forms.py (hunk 12 above)
Comment (mistercrunch) [inline superset/forms.py]:
```
In [1]: range(9)
Out[1]: [0, 1, 2, 3, 4, 5, 6, 7, 8]
```

File: superset/viz.py [hunk 14]
```diff
@@ -1208,14 +1208,21 @@ class DistributionPieViz(NVD3Viz):
...
+        'label': 'Chart Options',
+        'fields': (
             'pie_label_type',
             ('donut', 'show_legend'),
-            'labels_outside',
-        )
+            ('labels_outside', 'pie_half_circle'),
+            ('pie_pad_angle', 'pie_corner_radius'),
+        ),
+        "description": _("Properties setting of default viewport.")
```
Comment (mistercrunch) [inline superset/viz.py]: `_("Rendering options for the pie chart")`

File: superset/viz.py (hunk 13 above)
Comment (mistercrunch) [inline superset/viz.py]: agreed, better `None` than `'Criterion'`

**Review by mistercrunch — CHANGES_REQUESTED** [review]
Would love to see screenshots, I'm curious to see whether the half-pie centers properly and how the other options look like (trying to understand how that may be preferable to existing options).

Rounding the corners may not be visually accurate as small charts get more shaved off proportionally than large ones. Not a big deal, but styling should not diminish visual accuracy.

Also to note is that you currently need to do your change in exploreV2 as well until we fully migrate there. Essentially the configuration elements in `viz.py` and `forms.py` are moving to the javascript layer. Let us know if you can't find the place to change the file and the endpoint for exploreV2 to test it

File: superset/models.py (hunk 11 above)
Comment (ganshanshan) [inline superset/models.py]: Sorry,i remove it.this is for jinja {{}} in the SQL ,no only in WHEWE or HAVING.

File: superset/viz.py (hunk 14 above)
Comment (ganshanshan) [inline superset/viz.py]: Thanks very much, I update my code.

Comment (ganshanshan) [inline superset/viz.py]: To have your reply, it's my pleasure.

1.The screenshots 
You can find the screenshots in #1684 .

2.I update my code.please have a check again.

3.I do not know how to change the files in exploreV2,and do not know how to test it in V2.
I issue this before #1677.

Look forward to your reply.

**Comment by mistercrunch** [comment]
There's been merge conflicts for a long time, feel free to re-open once conflicts are sorted out.
